load_tests/benchmarks/k6.py

```diff
@@ -0,0 +1,233 @@
+        # check if the file already exists
+        if not os.path.exists("inputs_constant_tokens.json"):
+            logger.info(f'Preparing input file with {self.input_tokens_len} input tokens')
+            outputs = []
+            with open(self.k6_config.conversation_input_file, "r") as f:
+                data = json.load(f)
+                for doc in data:
+                    for conversation in doc["conversations"]:
+                        if not conversation["from"] == "human":
+                            continue
+                        if get_tokens_count(conversation["value"]) < self.input_tokens_len:
+                            continue
+                        # encode the message
+                        encoding = self.k6_config.tokenizer(conversation["value"], truncation=True,
+                                                            max_length=self.input_tokens_len)
+                        # find last encoded characters
+                        span = encoding.token_to_chars(len(encoding["input_ids"]) - 1)
+                        outputs.append(
+                            {"message": conversation["value"][0:span.end], "num_tokens": len(encoding["input_ids"])})
+                    if len(outputs) >= MAX_SAMPLES:  # limit the number of inputs
+                        break
+            with open("inputs_constant_tokens.json", "w") as f:
+                f.write(json.dumps(outputs))
+
+        # create a second input file with a sampling of inputs
+        # check if the file already exists
+        if not os.path.exists("inputs_variable_tokens.json"):
+            logger.info(
+                f'Preparing input file by randomly sampling shareGPT conversations at "{self.k6_config.conversation_input_file}"')
+            outputs = []
+            with open(self.k6_config.conversation_input_file, "r") as f:
+                data = json.load(f)
+                num_docs = len(data)
+                # generate random indexes to sample the data
+                indexes = np.random.choice(num_docs, 200, replace=False)
+                for i in indexes:
+                    doc = data[i]
+                    for conversation in doc["conversations"]:
+                        if not conversation["from"] == "human":
+                            continue
+                        # encode the message without truncation
+                        encoding = self.k6_config.tokenizer(conversation["value"])
+                        outputs.append(
+                            {"message": conversation["value"], "num_tokens": len(encoding["input_ids"])})
+                    if len(outputs) >= MAX_SAMPLES:  # limit the number of inputs
+                        break
+            with open("inputs_variable_tokens.json", "w") as f:
+                f.write(json.dumps(outputs))
+
+    def run(self):
+        self.k6_config.executor.render()
+        args = f"/tmp/k6-sse run --out json=results.json {self.k6_config.executor.rendered_file}"
```

Comment:
Hey there, awesome work @Hugoch !

Following up on @ErikKaum comment, why not use vanilla k6 and  gather Time To First Token, Time Per Output Token, etc... from response headers?

This is a snippet from my own TGI load testing tool, maybe it is useful:
```
        // Extract per request stats
        const end_to_end_time = res.timings.duration;
        const generated_tokens = parseInt(res.headers['X-Generated-Tokens']);
        const inference_time = parseInt(res.headers['X-Inference-Time']);
        const prompt_tokens = parseInt(res.headers['X-Prompt-Tokens']);
        const queue_time = parseInt(res.headers['X-Queue-Time']);
        const time_per_output_token = parseInt(res.headers['X-Time-Per-Token']);
        const total_time = parseInt(res.headers['X-Total-Time']);
        const validation_time = parseInt(res.headers['X-Validation-Time']);

        // Compute overhead time from I/0
        const overhead_time = end_to_end_time - total_time
        // Compute time to first token
        const time_to_first_token = inference_time - generated_tokens * time_per_output_token
        ```

Reply:
Hey @martinigoyanes ! Thx for the feedback!
Headers can be used but do not precisely account for the whole e2e time with HTTP layer. Time to first token cannot be precisely measured: with  `const time_to_first_token = inference_time - generated_tokens * time_per_output_token` as you assume that time per token is the same for all tokens. It is true in average for all n>1 tokens, but you are missing the prefill phase which will delay first token.